### PR TITLE
[CALCITE-2157] ClickHouse dialect implementation.

### PR DESCRIPTION
**Pull request by chris-baynes**

### CODE REVIEW

**Comment by suez1224** [comment]
Also, please rebase your branch.

**Comment by julianhyde** [comment]
We don't usually put a SQL database on the powered-by page when we add its dialect to the JDBC adapter. Maybe it seems "too easy".

Maybe should add a list of supported dialects to https://calcite.apache.org/docs/adapter.html ?

**Comment by julianhyde** [comment]
A section, or maybe a paragraph. I don't want it to sound too official, because we don't have the resources to properly test each dialect every release.

Does this change come with a VM or other means of testing against Clickhouse? It isn't absolutely necessary but it is preferred.

You are correct, Calcite's parser only supports its own dialect. We have added quite a few optional extensions, so that dialect is fairly forgiving. When we say "support a dialect" we are talking about what the JDBC adapter can generate. It needs to know the capabilities of the target database (e.g. whether it supports window functions) and syntax features (e.g. the character for quoting identifiers).

**Comment by YangChuanMakeSense** [comment]
@chris-baynes hello baynes, My work is based on calcite, and I use clickhouse too. I really want to know the progress of clickhouse sql dialect. very appreciate of you for giving any information.

**Comment by chris-baynes** [comment]
@YangChuanMakeSense I've updated this PR, if you want to follow the discussion on getting it merged please take a look at the jira ticket: https://jira.apache.org/jira/browse/CALCITE-2157

**Comment by love1693294577** [comment]
Hello, how is your clickhouse used, how can I still not connect, it seems to be unable to load the driver


Caused by: java.lang.RuntimeException: java.sql.SQLException: Cannot load JDBC driver class 'ru.yandex.clickhouse.ClickHouseDriver'
	at org.apache.calcite.adapter.jdbc.JdbcUtils$DialectPool.get(JdbcUtils.java:97)
	at org.apache.calcite.adapter.jdbc.JdbcSchema.createDialect(JdbcSchema.java:197)
	at org.apache.calcite.adapter.jdbc.JdbcSchema.create(JdbcSchema.java:133)
	at org.apache.calcite.adapter.jdbc.JdbcSchema.create(JdbcSchema.java:120)
	at org.apache.calcite.adapter.jdbc.JdbcSchema.create(JdbcSchema.java:172)
	at org.apache.calcite.adapter.jdbc.JdbcSchema$Factory.create(JdbcSchema.java:560)
	at org.apache.calcite.model.ModelHandler.visit(ModelHandler.java:282)
	... 18 more
Caused by: java.sql.SQLException: Cannot load JDBC driver class 'ru.yandex.clickhouse.ClickHouseDriver'
	at org.apache.commons.dbcp2.BasicDataSource.createConnectionFactory(BasicDataSource.java:489)
	at org.apache.commons.dbcp2.BasicDataSource.createDataSource(BasicDataSource.java:599)
	at org.apache.commons.dbcp2.BasicDataSource.getConnection(BasicDataSource.java:809)
	at org.apache.calcite.adapter.jdbc.JdbcUtils$DialectPool.get(JdbcUtils.java:78)

**Comment by love1693294577** [comment]
Hello, how is your clickhouse used, how can I still not connect, it seems to be unable to load the driver


Caused by: java.lang.RuntimeException: java.sql.SQLException: Cannot load JDBC driver class 'ru.yandex.clickhouse.ClickHouseDriver'
at org.apache.calcite.adapter.jdbc.JdbcUtils$DialectPool.get(JdbcUtils.java:97)
at org.apache.calcite.adapter.jdbc.JdbcSchema.createDialect(JdbcSchema.java:197)
at org.apache.calcite.adapter.jdbc.JdbcSchema.create(JdbcSchema.java:133)
at org.apache.calcite.adapter.jdbc.JdbcSchema.create(JdbcSchema.java:120)
at org.apache.calcite.adapter.jdbc.JdbcSchema.create(JdbcSchema.java:172)
at org.apache.calcite.adapter.jdbc.JdbcSchema$Factory.create(JdbcSchema.java:560)
at org.apache.calcite.model.ModelHandler.visit(ModelHandler.java:282)
... 18 more
Caused by: java.sql.SQLException: Cannot load JDBC driver class 'ru.yandex.clickhouse.ClickHouseDriver'
at org.apache.commons.dbcp2.BasicDataSource.createConnectionFactory(BasicDataSource.java:489)
at org.apache.commons.dbcp2.BasicDataSource.createDataSource(BasicDataSource.java:599)
at org.apache.commons.dbcp2.BasicDataSource.getConnection(BasicDataSource.java:809)
at org.apache.calcite.adapter.jdbc.JdbcUtils$DialectPool.get(JdbcUtils.java:78)


My configuration：
{
  version: '1.0',
  defaultSchema: 'click',
  schemas: [
    {
      name: 'click',
      type: 'custom',
      factory: 'org.apache.calcite.adapter.jdbc.JdbcSchema$Factory',
      operand: {
        jdbcDriver: 'ru.yandex.clickhouse.ClickHouseDriver',
        jdbcUrl: 'jdbc:clickhouse://XX.XX.XX.XX:8123/default',
        jdbcUser: 'druid',
        jdbcPassword: 'diurd'
      }
    }
  ]
}

**Comment by love1693294577** [comment]
Hello, how is your clickhouse used, how can I still not connect, it seems to be unable to load the driver

Caused by: java.lang.RuntimeException: java.sql.SQLException: Cannot load JDBC driver class 'ru.yandex.clickhouse.ClickHouseDriver'
at org.apache.calcite.adapter.jdbc.JdbcUtils$DialectPool.get(JdbcUtils.java:97)
at org.apache.calcite.adapter.jdbc.JdbcSchema.createDialect(JdbcSchema.java:197)
at org.apache.calcite.adapter.jdbc.JdbcSchema.create(JdbcSchema.java:133)
at org.apache.calcite.adapter.jdbc.JdbcSchema.create(JdbcSchema.java:120)
at org.apache.calcite.adapter.jdbc.JdbcSchema.create(JdbcSchema.java:172)
at org.apache.calcite.adapter.jdbc.JdbcSchema$Factory.create(JdbcSchema.java:560)
at org.apache.calcite.model.ModelHandler.visit(ModelHandler.java:282)
... 18 more
Caused by: java.sql.SQLException: Cannot load JDBC driver class 'ru.yandex.clickhouse.ClickHouseDriver'
at org.apache.commons.dbcp2.BasicDataSource.createConnectionFactory(BasicDataSource.java:489)
at org.apache.commons.dbcp2.BasicDataSource.createDataSource(BasicDataSource.java:599)
at org.apache.commons.dbcp2.BasicDataSource.getConnection(BasicDataSource.java:809)
at org.apache.calcite.adapter.jdbc.JdbcUtils$DialectPool.get(JdbcUtils.java:78)

My configuration：
{
version: '1.0',
defaultSchema: 'click',
schemas: [
{
name: 'click',
type: 'custom',
factory: 'org.apache.calcite.adapter.jdbc.JdbcSchema$Factory',
operand: {
jdbcDriver: 'ru.yandex.clickhouse.ClickHouseDriver',
jdbcUrl: 'jdbc:clickhouse://XX.XX.XX.XX:8123/default',
jdbcUser: '',
jdbcPassword: ''
}
}
]
}

**Comment by Jack0000007** [comment]
> @YangChuanMakeSense I've updated this PR, if you want to follow the discussion on getting it merged please take a look at the jira ticket: https://jira.apache.org/jira/browse/CALCITE-2157

I think this PR is  uesful to me . When will calcite support more syntax of clickhouse such as join syntax ?And will this PR merged into the master branch soon ?

File: core/src/main/java/org/apache/calcite/sql/dialect/ClickHouseSqlDialect.java [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+
+    fetch.unparse(writer, -1, -1);
+    writer.endList(frame);
+  }
+
+  @Override public void unparseCall(SqlWriter writer, SqlCall call,
+      int leftPrec, int rightPrec) {
+    if (call.getOperator() == SqlStdOperatorTable.SUBSTRING) {
+      CLICKHOUSE_SUBSTRING.unparse(writer, call, 0, 0);
+    } else {
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/sql/dialect/ClickHouseSqlDialect.java:167]: Why we must have a `CLICKHOUSE_SUBSTRING ` operator there ?

File: core/src/main/java/org/apache/calcite/sql/dialect/ClickHouseSqlDialect.java [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+import org.apache.calcite.sql.parser.SqlParserPos;
+
+import com.google.common.base.Preconditions;
+
+/**
+ * A <code>SqlDialect</code> implementation for the ClickHouse database.
+ */
+public class ClickHouseSqlDialect extends SqlDialect {
+  public static final SqlDialect DEFAULT =
+      new ClickHouseSqlDialect(EMPTY_CONTEXT
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/sql/dialect/ClickHouseSqlDialect.java:45]: We may also need a `DEFAULT_CONTEXT` for extensibility, see CALCITE-3416 for details.

File: core/src/main/java/org/apache/calcite/sql/dialect/ClickHouseSqlDialect.java (hunk 1 above)
Comment (chris-baynes) [reply core/src/main/java/org/apache/calcite/sql/dialect/ClickHouseSqlDialect.java:167]: The clickhouse substring function looks like this: `substring(s, offset, length)` which is a bit different from the calcite `SUBSTRING(string FROM integer FOR integer)`

File: core/src/main/java/org/apache/calcite/sql/dialect/ClickHouseSqlDialect.java (hunk 2 above)
Comment (chris-baynes) [reply core/src/main/java/org/apache/calcite/sql/dialect/ClickHouseSqlDialect.java:45]: I'll make another pass at a rebase on this branch soon and will fix it. Thanks for the review, would love to get this merged!

File: core/src/main/java/org/apache/calcite/sql/dialect/ClickHouseSqlDialect.java (hunk 1 above)
Comment (danny0405) [reply core/src/main/java/org/apache/calcite/sql/dialect/ClickHouseSqlDialect.java:167]: I mean just to unparse with CLICKHOUSE style when we encounter the `SqlStdOperatorTable.SUBSTRING`, seems not necessary with an operator there.

Comment (chris-baynes) [reply core/src/main/java/org/apache/calcite/sql/dialect/ClickHouseSqlDialect.java:167]: It's just a way to keep that logic in one place. But I'm not opposed to calling the writer directly here if you think it's better.

File: core/src/main/java/org/apache/calcite/sql/dialect/ClickHouseSqlDialect.java (hunk 2 above)
Comment (chris-baynes) [reply core/src/main/java/org/apache/calcite/sql/dialect/ClickHouseSqlDialect.java]: @danny0405 fixed this and some other things in the rebase.

**Comment by chris-baynes** [comment]
@danny0405 check style is fixed

**Comment by pilife** [comment]
@chris-baynes Hi, is there any future plan to support the join syntax?

**Pull request closed without merging**